qubesadmin/tools/qvm_template.py

```diff
@@ -0,0 +1,802 @@
+                    str(datetime.datetime.fromtimestamp(
+                        int(package_hdr[rpm.RPMTAG_BUILDTIME])))
+                tpl.features['template-install-time'] = \
+                    str(datetime.datetime.today())
+                tpl.features['template-license'] = \
+                    package_hdr[rpm.RPMTAG_LICENSE]
+                tpl.features['template-url'] = \
+                    package_hdr[rpm.RPMTAG_URL]
+                tpl.features['template-summary'] = \
+                    package_hdr[rpm.RPMTAG_SUMMARY]
+                tpl.features['template-description'] = \
+                    package_hdr[rpm.RPMTAG_DESCRIPTION].replace('\n', '|')
+    finally:
+        os.remove(LOCK_FILE)
+
+def qrexec_popen(args, app, service, stdout=subprocess.PIPE, filter_esc=True):
+    if args.updatevm:
+        return app.domains[args.updatevm].run_service(
+            service,
+            filter_esc=filter_esc,
+            stdout=stdout)
+    return subprocess.Popen([
+            '/etc/qubes-rpc/%s' % service,
+        ],
+        stdin=subprocess.PIPE,
+        stdout=stdout,
+        stderr=subprocess.PIPE)
+
+def qrexec_payload(args, app, spec, refresh):
+    # TODO: Support for force-refresh
+    _ = app # unused
+
+    def check_newline(string, name):
+        if '\n' in string:
+            parser.error(f"Malformed {name}:" +
+                " argument should not contain '\\n'.")
+
+    payload = ''
+    for repo in args.enablerepo if args.enablerepo else []:
+        check_newline(repo, '--enablerepo')
+        payload += '--enablerepo=%s\n' % repo
+    for repo in args.disablerepo if args.disablerepo else []:
+        check_newline(repo, '--disablerepo')
+        payload += '--disablerepo=%s\n' % repo
+    for repo in args.repoid if args.repoid else []:
+        check_newline(repo, '--repoid')
+        payload += '--repoid=%s\n' % repo
+    if refresh:
+        payload += '--refresh\n'
+    check_newline(args.releasever, '--releasever')
+    payload += '--releasever=%s\n' % args.releasever
+    check_newline(spec, 'template name')
+    payload += spec + '\n'
+    payload += '---\n'
+    for path in args.repo_files:
+        with open(path, 'r') as fd:
+            payload += fd.read() + '\n'
+    return payload
+
+def qrexec_repoquery(args, app, spec='*', refresh=False):
+    proc = qrexec_popen(args, app, 'qubes.TemplateSearch')
+    payload = qrexec_payload(args, app, spec, refresh)
+    stdout, stderr = proc.communicate(payload.encode('UTF-8'))
+    stdout = stdout.decode('ASCII')
+    if proc.wait() != 0:
+        for line in stderr.decode('ASCII').rstrip().split('\n'):
+            print('[Qrexec] %s' % line, file=sys.stderr)
+        raise ConnectionError("qrexec call 'qubes.TemplateSearch' failed.")
+    name_re = re.compile(r'^[A-Za-z0-9._+\-]*$')
+    evr_re = re.compile(r'^[A-Za-z0-9._+~]*$')
+    date_re = re.compile(r'^\d+-\d+-\d+ \d+:\d+$')
+    licence_re = re.compile(r'^[A-Za-z0-9._+\-()]*$')
```

Comment:
`-` must be the last in the [] set, otherwise it is interpreted as a chars range.

Reply:
Both in my testing and from regex101, the backslash before the `-` seem to work as intended?

Reply:
Ah, indeed, I haven't noticed backslash.